### PR TITLE
Homework 2.

### PR DESCRIPTION
**Pull request by interist32**

### CODE REVIEW

File: hw-2/controllers/users.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+const usersStorage = new UsersStorage();
+
+function prepareUser(user: User | null): User | null {
+  if (user === null) return null;
+
+  const preparedUser = {
+    ...user,
+  };
+  delete preparedUser.isDeleted;
```
Comment (aledpardo) [inline hw-2/controllers/users.ts]: As a tip:

There's an `ES2015` way of "deleting" a prop:

```typescript
const { isDeleted: _, ...user: preparedUser } = user;
```

There's a V8 blog post where they kinda say that using `delete` indiscriminately might degrade performance issues... Couldn't find it to show you...

File: hw-2/controllers/users.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+  return preparedUser;
+}
+
+export const getUsers = (req: Request, res: Response): void => {
```
Comment (aledpardo) [inline hw-2/controllers/users.ts]: We may improve readability a bit:

```typescript
export const getUsers = ({ query }: Request, res: Response): void => {
  const { login, limit: limitQs } = query;
  const limit = +limit || undefined;
  const users = usersStorage.getUsers(login, limit).map(prepareUser);

  res.json(users);
};
```

File: hw-2/controllers/users.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  const users = usersStorage.getUsers(
+    req.query.login, Number(req.query.limit) || undefined,
+  ).map((user) => prepareUser(user));
+  res.json(users);
+};
+
+export const getUser = (req: Request, res: Response): void => {
+  res.json(prepareUser(usersStorage.getUserById(req.params.id)));
```
Comment (aledpardo) [inline hw-2/controllers/users.ts]: When we create intermediary variables, the code is easier to read:

```typescript
export const getUser = ({ params }: Request, res: Response): void => {
  const user = usersStorage.getUserById(params.id);
  const preparedUser = prepareUser(user);

  res.json(preparedUser);
};
```

File: hw-2/models/user.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
+import * as Joi from '@hapi/joi';
+import { ContainerTypes, ValidatedRequestSchema } from 'express-joi-validation';
+
+/** User model. */
+export interface User {
+  id: string;
+  login: string;
+  password: string;
+  age: number;
+  isDeleted: boolean;
```
Comment (aledpardo) [inline hw-2/models/user.ts]: We could make `isDeleted` to be optional: `isDeleted?: boolean`

File: hw-2/storage/users.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+      users = users.slice(0, limit);
+    }
+    return users;
+  }
+
+  update(user: User): User | null {
+    const existingUser = this.getUserById(user.id);
+    if (!existingUser) return null;
+
+    const updatedUser = {
```
Comment (aledpardo) [inline hw-2/storage/users.ts]: Here we could simplify a little bit:

```typescript
const { id, isDeleted, ...existingUser } = this.getUserById(user.id) || {};

if (!id) return null;

const updatedUser = {
    ...existingUser,
    ...user,
    isDeleted,
    id
};

File: hw-2/tsconfig.json [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+    /* Additional Checks */
+    // "noUnusedLocals": true,                /* Report errors on unused locals. */
+    // "noUnusedParameters": true,            /* Report errors on unused parameters. */
+    // "noImplicitReturns": true,             /* Report error when not all code paths in function return a value. */
+    // "noFallthroughCasesInSwitch": true,    /* Report errors for fallthrough cases in switch statement. */
+
+    /* Module Resolution Options */
+    // "moduleResolution": "node",            /* Specify module resolution strategy: 'node' (Node.js) or 'classic' (TypeScript pre-1.6). */
+    // "baseUrl": "./",                       /* Base directory to resolve non-absolute module names. */
```
Comment (aledpardo) [inline hw-2/tsconfig.json]: Here's a challenge for you:

To combine the `baseUrl` and `paths` to configure the import paths so this:

```typescript
import { User } from '../models/user';
```

Can be replaced by:

```typescript
import { User } from '@app-models/user';
```


Do you accept the challenge? If yes, you can try it out on the next homework. ;)

File: hw-2/controllers/users.ts (hunk 1 above)
Comment (interist32) [reply hw-2/controllers/users.ts]: Wow, nice tip. Was thinking how to make with destructuring. Thanks you!

File: hw-2/controllers/users.ts (hunk 2 above)
Comment (interist32) [reply hw-2/controllers/users.ts]: Nice! I believe it should be 
```
...
const limit = +limitQs || undefined;
...
```
Fixed!

File: hw-2/models/user.ts (hunk 4 above)
Comment (interist32) [reply hw-2/models/user.ts]: We even have to! Thanks!

File: hw-2/controllers/users.ts (hunk 3 above)
Comment (interist32) [reply hw-2/controllers/users.ts]: Got it! Thank you again!

File: hw-2/storage/users.ts (hunk 5 above)
Comment (interist32) [reply hw-2/storage/users.ts]: Done!

File: hw-2/tsconfig.json (hunk 6 above)
Comment (interist32) [reply hw-2/tsconfig.json]: Hm... Sounds interesting! I do. Will do :) Thanks!

**Pull request merged**